### PR TITLE
feat(visibility): Clamp date range for `TagStore` queries

### PR DESCRIPTION
**Pull request by gggritso**
Follow-up and complement to https://github.com/getsentry/sentry/pull/80332. In short, `SAMPLE` wasn't enough. For longer ranges (i.e., >30d) fetching a list of project tags still times out.

In this PR, we're adding time range clamping for fetching tag keys. We will only ever query a maximum of N days (14 for now, but it's configurable). If someone chooses to get tags for the last 90 days, they will get tags from the last 14 days. If they need tags from November 5th - November 10th, we will fetch the range they asked for.

We think this is a reasonable compromise. When tags don't load at all, autocomplete stops working, which is very bad UX. Limiting the date range to 14 days is a tradeoff. The tags will succeed more often, but some tags might be missing.

Also, limiting the range should improve the cache hit ratio a bit, but we'll see.

### CODE REVIEW

File: src/sentry/api/utils.py [hunk 1]
```diff
@@ -234,6 +234,35 @@ def get_date_range_from_stats_period(
...
 
+def clamp_date_range(
+    range: tuple[datetime.datetime, datetime.datetime], max_timedelta: datetime.timedelta
+) -> tuple[datetime.datetime, datetime.datetime]:
+    """
+    Accepts a date range and a maximum time delta. If the date range is shorter
+    than the max delta, returns the range as-is. If the date range is longer than the max delta, clamps the range range, anchoring to the end.
+
+    If any of the inputs are invalid (e.g., a negative range) returns the range
+     without modifying it.
```
Comment (Zylphrex) [inline src/sentry/api/utils.py:245]:
```suggestion
    without modifying it.
```

File: src/sentry/api/utils.py [hunk 2]
```diff
@@ -234,6 +234,35 @@ def get_date_range_from_stats_period(
...
+
+    # Ignore negative ranges
+    if delta <= datetime.timedelta(0):
+        return (start, end)
+
+    # Ignore negative max time deltas
+    if max_timedelta < datetime.timedelta(0):
+        return (start, end)
+
+    return (end - min(delta, max_timedelta), end)
```
Comment (Zylphrex) [inline src/sentry/api/utils.py:263]: Nit: I think you can clean this up a little bit by checking `if delta < max_delta`
```suggestion
    # Ignore negative max time deltas
    if max_timedelta < datetime.timedelta(0):
        return (start, end)
        
    # Ignore if delta is within acceptable range
    if delta < max_delta:
        return (start, end)

    return (end - max_timedelta, end)
```

File: tests/sentry/api/test_utils.py [hunk 3]
```diff
@@ -196,3 +197,29 @@ def test_handle_query_errors(self, mock_parse_error):
                     raise ex
             except Exception as e:
                 assert isinstance(e, (FooBarError, APIException))
+
+
+class ClampDateRangeTest(unittest.TestCase):
```
Comment (Zylphrex) [inline tests/sentry/api/test_utils.py:202]: What happens if `max_timedelta` is set to `0`?

**Review by Zylphrex — APPROVED** [review]

**Comment by gggritso** [comment]
@Zylphrex took your suggestions, thanks!

**Pull request merged**